### ISSUE TITLE
Improve error handling

### ISSUE THREAD

**Issue by alvaromorales** [infolab-csail/WikipediaBase#82]
WikipediaBase should handle errors better.

Some suggested error types:

* **bad-request-format**: Bad request, can’t parse lisp	
* **function-not-found**: KB method doesn’t exist
* **symbol-not-found**: Symbol doesn’t exist
* **attribute-not-found**: Attribute for (symbol, class) doesn’t exist
* **class-not-found**: Class for symbol doesn’t exist
* **attribute-value-not-found**: Value doesn't exist, e.g. "currently alive"

EDIT: fixed cut&paste error

**Comment by sfelshin** [comment]
That was a cut-paste error above, right?  I think the second-to-last item was supposed to read

 * **class-not-found**: Class for symbol doesn’t exist

**Comment by sfelshin** [comment]
`get` should always return a list of results, of which each element should be a list (consisting of alternating typecodes and values), but other WikiBase commands, such as `get-classes` and `sort-symbols-named`, can return other types of results. Should we signal errors the same way for all commands?  If so, then `get` should probably return a simple list in order to signal an error, e.g.,

    (get "wikipedia-term" "Black Sea" (:code "non-existent attribute"))
    (:error attribute-not-found (:code "non-existent attribute"))

**Comment by chsamlee** [comment]
We probably should have something like illegal-argument for well-formed queries that have bad arguments (e.g. does not have :code or :render in attributes), and reserve bad-request-format for ill-formed queries (e.g. missing parentheses).

**Comment by sfelshin** [comment]
Good thinking. I agree, but rather than `illegal-argument`, I prefer something like `illegal-request` or `malformed-request` or `unrecognized-request` to make it clearer that the problem is with the whole request rather than some argument within the request.

**Comment by alvaromorales** [comment]
So far the conversation has been about increasing the number of errors we handle (and returning descriptive error types). In addition to that, we need to revamp our error handling mechanism.

We should raise exceptions instead of returning literal error objects (Golang-style). Passing error objects around is awkward in terms of control flow, and incompatible with Python's exception-raising model. Methods have to inspect return values (via `is_error`), instead of _handling_ errors in a `try/except` statement.

We should define a new module, `exceptions.py`, that has a base `WikipediaBaseException` class that accepts kwargs (e.g. `message`, `reply`). We can then have subclasses like `AttributeValueNotFound` for signaling more specific errors.

To lispify a Python exception, we'll just need to transform a `CamelCasedExceptionClassName` to a `hyphen-separated-string`. This should be added to `lispify.py`. (We'll no longer need to keep a dict of exceptions --> lispified names).

@sfelshin brought this up a few months ago:

>  Should we signal errors the same way for all commands?  If so, then `get` should probably return a simple list in order to signal an error, e.g.,

>    (get "wikipedia-term" "Black Sea" (:code "non-existent attribute"))
>    (:error attribute-not-found (:message "non-existent attribute"))

Currently, every method in Wikibase returns a list of plists (for the example above, we would return `((:error attribute-not-found (:message "non-existent attribute"))`. If we're only returning one exception, should we re-evaluate this?

@chsamlee brought up another interesting point: "[the output of `get`] may contain multiple errors or mix valid responses with an error. We should make sure that only the error is returned if the answers contain an error."

I can't think of a use case where we would want to mix answers with errors. If there is such a use case, we should discuss if it's worth the additional code complexity.

**Comment by sfelshin** [comment]
Here's a possible case where we might want to mix answers with errors:  Suppose we're fetching a rendered attribute that maps to multiple code attributes, and we get an error fetching one of them.  Currently that should be impossible, because START always starts by calling `get-attributes` on a specific article name (symbol), and `get-attributes` only returns names of attributes that can be successfully retrieved for the article.  But here are some ways that could change:
   * We decide to cache attributes, and the infobox for an article changes between when we cache the attributes and when we retrieve them.
   * For reasons of efficiency or whatever, we decide to have `get-attributes`, or some variant on `get-attributes`, return all attributes that are valid for the infobox rather than all attributes that are valid for the article.

Here's another possible case:  We fetch a date attribute for which we expect to be able to return both `:html` and `:yyyymmdd` values.  We're unable to parse a yyyymmdd value.  Currently we omit the yyyymmdd value in such cases, but we might decide it's helpful to return an error message indicating why.

My examples are hypothetical.  I agree with @alvaromorales: If and when we find a use case that we are actually interested in implementing, _then_ we should discuss if it's worth the additional code complexity.

**Comment by alvaromorales** [comment]
I think we first need to clarify what we mean by "an error":

* Unhandled runtime exception (e.g. undeclared variable, method called on a null object). **We can't recover from this** (it's _unhandled_; there is no error-handling code after all). We'll return an error object with the Python exception name (e.g. `NameError` --> `((:error "name-error" :message "name 'foo' is not defined"))`). 

  These kinds of errors should not happen. When they do, we need to push a bug fix. We don't want to have a generic `try / except` statement that catches any runtime error: if we do, we'll never surface corner cases and fix poorly written code.

* Expected error cases, like `attribute-value-not-found`, `class-not-found`, etc. We have error-handling code for these cases.

With these definitions, let's analyze each proposed use case and show that the suggested error models are not possible:

>  Suppose we're fetching a rendered attribute that maps to multiple code attributes

This use case is supported by the current implementation of the `get` method:

```
(get "wikipedia-class" "Article" (:rendered "Attribute"))
```

The specification for this method is: give me all values that have the given rendered attribute `x`.

The implementation sketch of this method is: we parse HTML to get a list of `(rendered, value)` pairs. If `rendered == x`, we append it to the result list. We then parse markup _from the DB_ to get a list of `(code, markup_value)` pairs. If `rendered_mappings[code] == x`, we append `markup_value` to the result list (we render `markup_value` to HTML if the `:markup "render"` option is set).

Back to the use case:

> ... we're fetching a rendered attribute that maps to multiple code attributes, **and we get an error fetching one of them**. 

Here's the hypothetical reason for this error:

> ... the infobox for an article changes between when we cache the attributes and when we retrieve them.

Let's say that `x` was removed from the article HTML. If so, we won't find it in the HTML, but we will find it in the markup in the DB.

What if START cached attributes a very long time ago, and we updated the WikipediaBase backend sometime after the attributes were cached? If so, `x` is no longer in the HTML **or** in the markup in the DB. We won't find `x` at all; we'll return an `attribute-value-not-found` error.

With these caching-related inconsistencies, there is no error model where we would mix answers with errors.

> For reasons of efficiency or whatever, we decide to have get-attributes, or some variant on get-attributes, return all attributes that are valid for the infobox rather than all attributes that are valid for the article.

START would cycle through each attribute `x` returned by the hypothetical `get-all-attributes` method, and call `get`. This is again handled by the implementation of `get`. We'll either find `x`, or we won't find it and will return `attribute-value-not-found`. There's no case in between these two outcomes.

> Here's another possible case: We fetch a date attribute for which we expect to be able to return both :html and :yyyymmdd values. We're unable to parse a yyyymmdd value. Currently we omit the yyyymmdd value in such cases, but we might decide it's helpful to return an error message indicating why.

It would be useful to surface the error message. But I don't think we want to return the error message in a separate _value_:

```
((:html "June 27, 1991") (:error "cant-parse-date" :message "computers are hard"))
```

This return format increases complexity, if we consider the case where we return multiple values:

```
((:html "June 27, 1991") (:error "cant-parse-date" :message "computers are hard") (:html "June 1, 2000") (:error "cant-parse-date" :message "foobar"))
```

With this model, `get` no longer returns a list of values, it returns a list of values, each followed by one or more optional error objects. This isn't a very clean return format.

We should instead return:

```
((:html "June 27, 1991" :warning "could not parse date to yyyymmdd: computers are hard"))
```

In this return format, the error message is scoped within the value that generated the error. `get` still returns a list of values.
